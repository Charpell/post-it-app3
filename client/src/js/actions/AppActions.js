import AppDispatcher from '../dispatcher/AppDispatcher';
import AppConstants from '../constants/AppConstants';


const AppActions = {
/**
   * @description describes an action that makes
   * API call to the server for a post request
   * to register a user
   *
   * @param {Object} userDetails an object that contains the username,
   * password, email and phone number of a new user
   *
   * @returns {void} void
   *
   */
  registerUser(userDetails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SIGN_UP,
      userDetails
    });
  },

/**
   * @description describes an action that makes
   * API call to the server for a post request
   * to receive a user details
   *
   * @param {Object} users an object containing the user details
   *
   *
   * @returns {void} void
   *
   * @returns {Object} returns registered user registration details
   */
  receiveUsers(users) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_USERS,
      users
    });
  },


/**
   * @description describes an action that makes
   * API call to the server for a post request to get all users
   * who have seen a message.
   *
   * @param {Object} user contains the user and the message Id of a
   * viewed message.
   *
   * @returns {Object} returns an object containing list of users
   */
  seenMessage(user) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SEEN_MESSAGE,
      user
    });
  },

/**
   * @description describes an action that makes
   *  API call to the server for a post request to save a user's group
   *
   * @param {Object} group contains the groupname and the user
   *
   * @returns {Object} returns notification message
   */
  createGroup(group) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.CREATE_GROUP,
      group
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all the groups that a user belongs to.
   *
   * @param {Object} userName the user
   *
   * @returns {void} void
   */
  getGroups(userName) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.GET_GROUPS,
      userName
    });
  },


  /**
   * @description describes an action that
   * receive all user's notification.
   *
   * @param {Object} notification an object of notifications
   *
   * @returns {Object} returns an object containing user's notification
   */
  receiveNotification(notification) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_NOTIFICATION,
      notification
    });
  },

/**
   * @description describes an action that
   * receive all groups that a user belong to.
   *
   * @param {Object} groups an object of groups
   *
   * @returns {Object} returns an object containing user's group
   */
  receiveGroups(groups) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_GROUPS,
      groups
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to save a user into a group.
   *
   * @param {Object} userDetails the groupname and the username
   *
   * @returns {Object} returns a notification message
   */
  addUserToGroup(userDetails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.ADD_USER_TO_GROUP,
      userDetails
    });
  },

  /**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to save a message.
   *
   * @param {Object} message an object that contains the group, message,
   * time it was posted, the user who posted the message
   *
   * @returns {void} void
   */
  postMessage(message) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.POST_MESSAGE,
      message
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to receive message.
   *
   * @param {Object} message an object of messages
   *
   * @returns {void} void
   */
  receiveMessages(message) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_MESSAGE,
      message
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to login a user.
   *
   * @param {Object} userDetails an object that contains the email and password
   * of a registered user
   *
   * @returns {void} void
   */
  loginUser(userDetails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SIGN_IN,
      userDetails
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to recieve a user's details.
   *
   * @param {Object} user an object containing the user details
   *
   * @returns {Object} returns registered user details
   */
  receiveLogin(user) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_LOGIN,
      user
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to get users and messages in a group.
   *
   * @param {String} group group name
   *
   * @returns {Object} returns users and messages in a group
   */
  searchUserMessage(group) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SEARCH_USER_MESSAGE,
      group
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to login a user with a google account
   *
   * @param {Object} googleUser an object that contains the username,
   * email and phone number of a new user
   *
   * @returns {Object} returns user's google account details
   */
  googleLogin(googleUser) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.GOOGLE_LOGIN,
      googleUser
    });
  },

 /**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to register a user with a google account
   *
   * @param {Object} googleUser an object that contains the username,
   * email and phone number of a new user
   *
   * @returns {Object} returns user's google account details
   */
  googleSignup(googleUser) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.GOOGLE_SIGNUP,
      googleUser
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server to sign the user out
   *
   * @returns {Object} returns registered user details
   */
  logout() {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.LOGOUT
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all users.
   *
   * @param {Object} users object containing list of users
   *
   * @returns {Object} returns an object containing list of users
   */
  receiveUser(users) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_USER,
      users
    });
  },

  /**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all emails.
   *
   * @param {Object} emails object containing list of emails
   *
   * @returns {Object} returns an object containing list of emails
   */
  receiveEmails(emails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_EMAILS,
      emails
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all numbers.
   *
   * @param {Object} numbers object containing list of numbers
   *
   * @returns {Object} returns an object containing list of numbers
   */
  receiveNumber(numbers) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_NUMBERS,
      numbers
    });
  },

/**
   * @description describes an action that gets
   * the current user from local storage
   *
   * @param {String} displayName dislplay name of the user
   *
   * @returns { String } returns a displayName
   */
  displayName(displayName) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.DISPLAY_NAME,
      displayName
    });
  },

 /**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to reset a user's password.
   *
   * @param {String} email the email address of the user
   *
   * @returns {Object} returns a notification message
   */
  resetPassword(email) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RESET_PASSWORD,
      email
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get a user's notification.
   *
   * @param {String} userName the user's name
   *
   * @returns {Object} returns an object containing user's notificaions
   */
  getNotification(userName) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.NOTIFICATIONS,
      userName
    });
  },

};
export default AppActions;
