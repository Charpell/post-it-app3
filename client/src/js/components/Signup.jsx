import React, { Component } from 'react';
import toastr from 'toastr';

import AppActions from '../actions/AppActions';
import AppStore from '../stores/AppStore';
import { firebaseAuth, firebase, provider } from '../../../../server/config';
import { validateEmail } from '../helpers/validate.helper';



/**
 * @description Gets user data and persits with firebase
 * 
 * @param {object} props
 * 
 * @class Signup
 * 
 * @extends {Component}
 */
export default class Signup extends Component {
    constructor(props) {
        super(props);
        this.state = {
            contacts: AppStore.getContacts(),
            databaseUsers: AppStore.getdatabaseUsers(),
            emails: AppStore.getGroupEmails(),
            numbers: AppStore.getAllUsersNumber(),
            googleUser: AppStore.getGoogleSignup()

        };
        this.onChange = this.onChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }


    /**
     * @method componentDidMount
     * 
     * @description Adds an event Listener to the Store and fires when the component is fully mounted.
     * 
     * @return {void}
     * 
     * @memberof Signup
     */
    componentDidMount() {
        AppStore.addChangeListener(this.onChange);
    }

    /**
    * @method componentWillUnmount
    *
    * @description Removes event Listener from the Store
    *
    * @memberof Signup
    */
    componentWillUnmount() {
        AppStore.removeChangeListener(this.onChange);
    }


    /**
     * @method onChange
     * 
     * @description Monitors changes in the components and change the state
     * 
     * @memberof Signup
     */
    onChange() {
        this.setState({
            databaseUsers: AppStore.getdatabaseUsers(),
            emails: AppStore.getGroupEmails(),
            numbers: AppStore.getAllUsersNumber(),
            googleUser: AppStore.getGoogleSignup()
        });

    }

    
    /**
    * @description Makes an action call to Sign up a user with username, email, phone number  and password
    *
    * @param {object} event
    *
    * @returns {void}
    *
    * @memberof Signup
    */
    handleSubmit(event) {
        event.preventDefault();

        /**
    * @method function
    * 
    * @param {any} string
    *
    * @memberof Signup
    */
        function capitalizeFirstLetter(string) {
            return string.charAt(0).toUpperCase() + string.slice(1);
        }

        const userNameToUppercase = capitalizeFirstLetter(this.refs.username.value);

        const contact = {
            username: userNameToUppercase,
            email: this.refs.email.value.trim(),
            password: this.refs.password.value.trim(),
            number: this.refs.number.value.trim()
        }

        if (this.state.databaseUsers.includes(userNameToUppercase)) {
            toastr.error('The username already exist')
        } else if (this.state.numbers.includes(this.refs.number.value)) {
            toastr.error('The phone number already exist')
        } else if (this.refs.password.value !== this.refs.verifyPassword.value) {
            toastr.error('Password does not match')

        } else if (!validateEmail(this.refs.email.value)) {
            toastr.error('Invalid Email Address')
        } else {
            AppActions.saveContact(contact);
            this.refs.username.value = '';
            this.refs.email.value = '';
            this.refs.password.value = '';
            this.refs.number.value = '';
            this.refs.verifyPassword.value = '';
        }
    }

    /**
     * @method render
     * 
     * @description Render react component
     * 
     * @memberof Signup
     */
    render() {
        return (
            <div>
                <div className="container" >
                    <div className="row">
                        <div className="col-sm-3"></div>
                        <div className="col-md-6 col-sm-6 col-xs-12">
                            <h3>Sign Up</h3>
                            <form onSubmit={this.handleSubmit}>
                                <div className='form-group'>
                                    <input type="text" ref='username' 
                                    className='form-control' 
                                    placeholder='Username' required />
                                </div>
                                <div className='form-group'>
                                    <input type="text" ref='email' 
                                    className='form-control' placeholder='Email' 
                                    required />
                                </div>
                                <div className='form-group'>
                                    <input type="text" ref='number' 
                                    className='form-control' 
                                    placeholder='Phone Number: Ex 2348066098146' 
                                    pattern="[234][0-9]{12}" 
                                    title="It will contain only 13 numbers and must start with 234" 
                                    required />
                                </div>
                                <div className='form-group'>
                                    <input type="password" ref='password' 
                                    className='form-control' 
                                    placeholder='Password' 
                                    pattern="(?=.*\d).{6,}" 
                                    title="Must contain at least 6 characters and 1 number" 
                                    required />
                                </div>
                                <div className='form-group'>
                                    <input type="password" ref='verifyPassword' 
                                    className='form-control' 
                                    placeholder='Verify Password' 
                                    pattern="(?=.*\d).{6,}" required />
                                </div>
                                <button type='submit' 
                                className='btn btn-primary'>Submit</button>
                            </form>
                        </div>
                        <div className="col-sm-3"></div>
                    </div>
                </div>
            </div>

        )
    }

}
