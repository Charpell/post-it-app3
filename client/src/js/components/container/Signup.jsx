import React, { Component } from 'react';
import toastr from 'toastr';

import AppActions from '../../actions/AppActions';
import AppStore from '../../stores/AppStore';
import { validateEmail } from './../../helpers/utils';
import Input from '../presentation/Input';


/**
 * @description the signup components lets users register an account
 *
 * @class SignUp
 *
 * @extends {Component}
 */
export default class Signup extends Component {
  /**
  * @description Creates an instance of Signup.
  * bind methods and set initial state.
  *
  * @memberof Signup
  *
  * @param {object} props
  */
  constructor(props) {
    super(props);
    this.state = {
      contacts: [],
      databaseUsers: [],
      emails: [],
      numbers: [],
      googleUser: null,
      username: '',
      email: '',
      number: '',
      password: '',
      verifyPassword: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.onChange = this.onChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  /**
   * @method componentDidMount
   *
   * @description A react life cycle method that listens to change
   * from the store,then fires an action.
   *
   * @return {void} void
   *
   * @memberof Signup
   */
  componentDidMount() {
    AppStore.addChangeListener(this.onChange);
  }

  /**
  * @method componentWillUnmount
  *
  * @description Removes event Listener from the Store
  *
  * @return {void} void
  *
  * @memberof Signup
  */
  componentWillUnmount() {
    AppStore.removeChangeListener(this.onChange);
  }

  /**
   * @description: controls inputs state
  *
  * @param {Object} element the current element
  *
  * @return {void} void
  */
  handleChange(element) {
    this.setState({
      [element.target.name]: element.target.value
    });
  }

  /**
   * @method onChange
   *
   * @description this method gets data from the store and sets to state
   *
   * @returns {void} void
   *
   * @memberof Signup
   */
  onChange() {
    this.setState({
      databaseUsers: AppStore.getDatabaseUsers(),
      numbers: AppStore.getAllUsersNumber(),
      googleUser: AppStore.getGoogleSignup()
    });
  }


  /**
  * @description describes a method that makes an action call
  * to Sign up a user with username, email, phone number  and password
  *
  * @param {object} event the event object
  *
  * @returns {void}
  *
  * @memberof Signup
  */
  handleSubmit(event) {
    event.preventDefault();

    const userDetails = {
      userName: this.state.userName,
      email: this.state.email,
      password: this.state.password,
      number: this.state.number
    };

    if (this.state.databaseUsers.includes(this.state.userName)) {
      toastr.error('The username already exist');
    } else if (this.state.numbers.includes(this.state.number)) {
      toastr.error('The phone number already exist');
    } else if (this.state.password !== this.state.verifyPassword) {
      toastr.error('Password does not match');
    } else if (!validateEmail(this.state.email)) {
      toastr.error('Invalid Email Address');
    } else {
      AppActions.registerUser(userDetails);
    }
  }

  /**
   * @description Render react component
   *
   * @memberof Signup
   *
   * @return { jsx } rendered jsx element
   */
  render() {
    return (
      <div>
        <div className="container" >
          <div className="row">
            <div className="col-sm-3"></div>
            <div className="col-md-6 col-sm-6 col-xs-12">
              <h3>Sign Up</h3>
              <form onSubmit={this.handleSubmit}>
                <Input
                  name="userName"
                  type={'text'}
                  action={this.handleChange}
                  className={'form-control'}
                  placeholder={'Username'}
                />
                <Input
                  name="email"
                  type={'text'}
                  action={this.handleChange}
                  className={'form-control'}
                  placeholder={'Email'}
                />
                <div className='form-group'>
                  <input type="text" name='number'
                    onChange={this.handleChange}
                    className='form-control'
                    placeholder='Phone Number: Ex 2348066098146'
                    pattern="[234][0-9]{12}"
                    title="It will contain 13 numbers and must start with 234"
                    required
                  />
                </div>
                <div className='form-group'>
                  <input type="password" name='password'
                    onChange={this.handleChange}
                    className='form-control'
                    placeholder='Password'
                    pattern="(?=.*\d).{6,}"
                    title="Must contain at least 6 characters and 1 number"
                    required
                  />
                </div>
                <div className='form-group'>
                  <input
                    type="password"
                    name='verifyPassword'
                    onChange={this.handleChange}
                    className='form-control'
                    placeholder='Verify Password'
                    pattern="(?=.*\d).{6,}"
                    required
                  />
                </div>
                <button
                  type='submit'
                  className='btn btn-primary'>
                  Submit
                </button>
              </form>
            </div>
            <div className="col-sm-3"></div>
          </div>
        </div>
      </div>
    );
  }
}
