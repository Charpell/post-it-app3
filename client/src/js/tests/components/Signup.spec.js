import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Signup from '../../components/Signup'
import AppActions from '../../actions/AppActions'

jest.mock('../../../../../server/config', () => ({
  }));
jest.mock('../../actions/AppActions');


let spyOnDispatcher;
beforeEach(() => {
    spyOnDispatcher = spyOn(AppActions, 'saveContact');
});

afterEach(() => {
    spyOnDispatcher.mockReset();
});

describe('Signup Component', () => {
  it('Signup component should render as expected', () => {
    const tree = renderer.create(<Signup />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display the necessary elements', () => {
    const wrapper = mount(<Signup />);
    wrapper.instance().componentWillUnmount();
    expect(wrapper.find('div').length).toBe(7);
    expect(wrapper.find('h2').length).toBe(1);
    expect(wrapper.find('h4').length).toBe(1);
    expect(wrapper.find('form').length).toBe(1);
    expect(wrapper.find('input').length).toBe(1);
    expect(wrapper.find('a').length).toBe(1);
    expect(wrapper.find('br').length).toBe(3);
});

it('should expect saveContact Action to be called', () => {
    const event = {
        target: {
          name: 'name',
          value: 'value',
        },
        preventDefault: () => jest.fn()
      };
    const wrapper = mount(<Signup />);
    wrapper.instance().refs.email.value = 'someemail@email.com';
    wrapper.instance().handleSubmit(event);
    expect(spyOnDispatcher).toHaveBeenCalled();
});


});
