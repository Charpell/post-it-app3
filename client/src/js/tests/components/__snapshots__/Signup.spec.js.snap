// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Component About component should render as expected 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-3"
      />
      <div
        className="col-md-6 col-sm-6 col-xs-12"
      >
        <h3>
          Sign Up
        </h3>
        <form
          onSubmit={[Function]}
        >
          <div
            className="form-group"
          >
            <input
              className="form-control"
              placeholder="Username"
              required={true}
              type="text"
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              placeholder="Email"
              required={true}
              type="text"
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              pattern="[234][0-9]{12}"
              placeholder="Phone Number: Ex 2348066098146"
              required={true}
              title="It will contain only 13 numbers and must start with 234"
              type="text"
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              pattern="(?=.*\\\\d).{6,}"
              placeholder="Password"
              required={true}
              title="Must contain at least 6 characters and 1 number"
              type="password"
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              pattern="(?=.*\\\\d).{6,}"
              placeholder="Verify Password"
              required={true}
              type="password"
            />
          </div>
          <button
            className="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
      <div
        className="col-sm-3"
      />
    </div>
  </div>
</div>
`;
